judge/ratings.py

```diff
@@ -8,6 +8,46 @@
 from django.utils import timezone
 
 
+BETA2 = 328.33 ** 2
+MEAN_INIT = 1500.
+VAR_INIT = 350**2 * (BETA2 / 212**2)
+SD_INIT = sqrt(VAR_INIT)
+VALID_RANGE = MEAN_INIT - 20 * SD_INIT, MEAN_INIT + 20 * SD_INIT
+VAR_PER_CONTEST = 1219.047619 * (BETA2 / 212**2)
+VAR_LIM = (sqrt(VAR_PER_CONTEST**2 + 4 * BETA2 * VAR_PER_CONTEST) - VAR_PER_CONTEST) / 2
+SD_LIM = sqrt(VAR_LIM)
+TANH_C = sqrt(3) / pi
+
+
+def eval_tanhs(tanh_terms, x):
+    return sum((wt / sd) * tanh((x - mu) / (2 * sd)) for mu, sd, wt in tanh_terms)
+
+
+def solve(tanh_terms, y_tg, lin_factor=0, bounds=VALID_RANGE):
+    L, R = bounds
+    Ly, Ry = None, None
+    while R - L > 2:
+        x = (L + R) / 2
+        y = lin_factor * x + eval_tanhs(tanh_terms, x)
+        if y > y_tg:
+            R, Ry = x, y
+        elif y < y_tg:
+            L, Ly = x, y
+        else:
+            return x
+    # Use linear interpolation to be slightly more accurate.
+    if Ly is None:
+        Ly = lin_factor * L + eval_tanhs(tanh_terms, L)
+    if y_tg <= Ly:
+        return L
+    if Ry is None:
+        Ry = lin_factor * R + eval_tanhs(tanh_terms, R)
+    if y_tg >= Ry:
+        return R
+    ratio = (y_tg - Ly) / (Ry - Ly)
+    return L * (1 - ratio) + R * ratio
+
+
 def tie_ranker(iterable, key=attrgetter('points')):
```

Comment:
Nit: could we move this function above or below the functions of computing ratings? Right now, it's in the middle (unless there's a reason for doing this that I missed).

Reply:
I moved this function upwards.